static-analysis/teammates-macker.xml

```diff
@@ -1,6 +1,161 @@
+            </deny>
+        </access-rule>
+    </ruleset>
+    <ruleset name="packageDiagram: Logic should not touch UI">
+        <access-rule>
+            <deny>
+                <from class="teammates.logic.**" />
+                <to class="teammates.ui.**" />
+            </deny>
+        </access-rule>
+    </ruleset>
+    <ruleset name="packageDiagram: Storage should not touch Logic">
+        <access-rule>
+            <deny>
+                <from class="teammates.storage.**" />
+                <to class="teammates.logic.**" />
+            </deny>
+        </access-rule>
+    </ruleset>
+    <ruleset name="packageDiagram: Storage should not touch UI">
+        <access-rule>
+            <deny>
+                <from class="teammates.storage.**" />
+                <to class="teammates.ui.**" />
+            </deny>
+        </access-rule>
+    </ruleset>
+    <ruleset name="packageDiagram: Client should not touch UI">
+        <access-rule>
+            <deny>
+                <from class="teammates.client.**" />
+                <to class="teammates.ui.**" />
+            </deny>
+        </access-rule>
+    </ruleset>
+    <ruleset name="packageDiagram: client::remoteapi can be only access by client::scripts">
+        <access-rule>
+            <deny>
+                <to class="teammates.client.remoteapi.**" />
+                <allow>
+                    <from class="teammates.client.scripts.**" />
+                </allow>          
+            </deny>
+        </access-rule>
+    </ruleset>
+    <ruleset name="packageDiagram: client::util should only be used inside client">
+        <access-rule>
+            <deny>
+                <to class="teammates.client.scripts.util.**" />
+                <allow>
+                    <from class="teammates.client.**" />
+                </allow>          
+            </deny>
+        </access-rule>
+    </ruleset>
+    
+    <ruleset name="UiComponent: Only *Action in can touch logic api">
+        <access-rule>
+            <deny>
+                <from class="teammates.ui.**" />
+                <to class="teammates.logic.**" />   
+                <allow>
+                    <from>
+                        <include class="teammates.ui.controller.**" />
+                        <include class="teammates.ui.automated.**" />
+                    </from>
+                    <to class="teammates.logic.api.**" />
+                </allow>
+                <allow>
+                    <!-- to import sample data, need to be changed -->
+                    <from class="teammates.ui.controller.AdminInstructorAccountAddAction" />
+                    <to class="teammates.logic.backdoor.BackDoorLogic" />
+                </allow>    
+            </deny>
+        </access-rule>
+    </ruleset>
+    <ruleset name="UiComponent: *PageData is only used in *Action and ActionResult">
+        <access-rule>
+            <deny>
+                <to class="teammates.ui.teamplate.**" />
```

Comment:
By the way, `*PageData` is contained in the `ui.pagedata` package instead so the ruleset is kind of mistargeted.

The full ruleset for this should be:
- Allow `ui.controller` -> `ui.pagedata`
- Allow `ui.pagedata` -> `ui.template`
- Allow dependency within same package. Arguably this needs to be stricter, but let's do it another time.